glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,496 @@
+        );
+        schedule_internal(&glean, submitter, scheduler, fake_now);
+        assert_eq!(0, submitter_count.swap(0, Ordering::Relaxed));
+        assert_eq!(1, scheduler_count.swap(0, Ordering::Relaxed));
+    }
+
+    // `When` is responsible for date math. Let's make sure it's correct.
+    #[test]
+    fn test_when() {
+        let now = FixedOffset::east(0).ymd(2021, 4, 30).and_hms(15, 2, 10);
+        // `now` is after `SCHEDULED_HOUR` so should be zero:
+        assert_eq!(std::time::Duration::from_secs(0), When::Today.until(now));
+        // If we bring it back before `SCHEDULED_HOUR` it should give us the duration:
+        let earlier = now.date().and_hms(SCHEDULED_HOUR - 1, 0, 0);
+        assert_eq!(
+            std::time::Duration::from_secs(3600),
+            When::Today.until(earlier)
+        );
+
+        // `Tomorrow` and `Reschedule` should differ only in their `reason()`
+        assert_eq!(
+            std::time::Duration::from_secs(46670),
+            When::Tomorrow.until(now)
+        );
+        assert_eq!(
+            std::time::Duration::from_secs(46670),
+            When::Reschedule.until(now)
+        );
+        assert_eq!(When::Tomorrow.until(now), When::Reschedule.until(now));
+        assert_ne!(When::Tomorrow.reason(), When::Reschedule.reason());
+    }
+
+    // Scheduler tests mutate global state and thus must not be run in parallel.
+    // Otherwise one test could cancel the other.
+    // This Mutex aims to solve that.
+    static SCHEDULER_TEST_MUTEX: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));
+
+    // The scheduler has been designed to be cancellable. Can we cancel it?
+    #[test]
+    fn test_cancel() {
+        // First and foremost, all scheduler tests must ensure they start uncancelled.
+        // Perils of having shared state.
+        let _test_lock = SCHEDULER_TEST_MUTEX.lock().unwrap();
+        let (cancelled_lock, condvar) = &**TASK_CONDVAR; // One `*` for Lazy, the second for Arc
+        *cancelled_lock.lock().unwrap() = false;
+
+        // Pick a time at least two hours from the next scheduled submission.
+        let now = FixedOffset::east(0)
+            .ymd(2021, 4, 30)
+            .and_hms(SCHEDULED_HOUR - 2, 0, 0);
+
+        let proxy_factory = || {
+            new_proxies(
+                |_, reason| {
+                    panic!(
+                        "Shouldn't submit when testing scheduler. reason: {:?}",
+                        reason
+                    )
+                },
+                |_, _| panic!("Not even using the scheduler this time."),
+            )
+        };
+
+        // Test Today.
+        let (submitter, submitter_count, _, _) = proxy_factory();
+        let handle = start_scheduler(submitter, now, When::Today);
+        *cancelled_lock.lock().unwrap() = true; // Cancel the scheduler thread.
+        condvar.notify_all(); // Notify any/all listening schedulers to check whether they were cancelled.
```

Comment:
Shouldn't this just call `cancel()`? It's what the outside will call and does the same inside, so we don't need to repeat it over and over again in tests.

```suggestion
        super::cancel();
```